ls1intum/Apollon_standalone
# Support Version Management (without sharing modes)

## Description

This PR adds version management as described in #97, with two exceptions:
- Version management is disabled in sharing modes. (edit, give/see feedback, collaborate) A follow-up PR is planned after this one is merged.
- I proposed two approaches in #97 for the modification of the diagram datatype to support version management. I made a different modification in my implementation as it seemed to be easier to implement, more efficient and avoided breaking changes:
```
id: string;
title: string;
model: UMLModel;
lastUpdate: string;
versions: Diagram[]; (recursive)
description: string;
token: string;
```

This avoids breaking changes because no existing fields are removed and only new fields are added. A diagram with two versions can then for example look like this:

```
{
  "id": "722378dd-e549-4621-9405-fe7f884e0689",
  "title": "Addressed Feedback",
  "model": {...},
  "lastUpdate": "2024-11-16T13:31:29.135Z",
  "description": "Incorporated feedback from the group",
  "versions": [
    {
      "id": "722378dd-e549-4621-9405-fe7f884e0689",
      "title": "Initial Draft",
      "model": {...},
      "lastUpdate": "2024-11-16T13:31:52.849Z",
      "description": "Proposal for the system"
    },
    {
      "id": "722378dd-e549-4621-9405-fe7f884e0689",
      "title": "Addressed Feedback",
      "model": {...},
      "lastUpdate": "2024-11-16T13:32:22.695Z",
      "description": "Incorporated feedback from the group"
    }
  ],
  "token": "pUX0hHPj5UYJE3an7Xk5"
}

```

### Steps for Testing

1. Model a diagram
2. Publish a version by opening the sidebar and pressing the '+' button in the top right
3. Make further changes and publish another version
4. Make sure that it is possible to delete, edit, preview, and restore versions. The button for restoring a version is visible after previewing the respective version.
